src/python/pants/core/util_rules/external_tool.py

```diff
@@ -0,0 +1,248 @@
+# Copyright 2020 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from abc import abstractmethod
+from dataclasses import dataclass
+from pathlib import PurePath
+from typing import List
+
+from pants.base.build_environment import get_buildroot
+from pants.core.util_rules.archive import ExtractedDigest, MaybeExtractable
+from pants.engine.fs import Digest, DirectoryToMaterialize, Snapshot, UrlToFetch
+from pants.engine.platform import Platform, PlatformConstraint
+from pants.engine.rules import RootRule, rule
+from pants.engine.selectors import Get
+from pants.subsystem.subsystem import Subsystem
+from pants.util.memo import memoized_method
+from pants.util.meta import classproperty
+
+
+class UnknownVersion(Exception):
+    pass
+
+
+class ExternalToolError(Exception):
+    pass
+
+
+@dataclass(frozen=True)
+class ExternalToolRequest:
+    url_to_fetch: UrlToFetch
+    exe: str
+
+
+@dataclass(frozen=True)
+class DownloadedExternalTool:
+    digest: Digest
+    exe: str
+
+
+class ExternalTool(Subsystem):
+    """Configuration for an invocable tool that we download from an external source.
+
+    Subclass this to configure a specific tool.
+
+
+    Idiomatic use:
+
+    class MyExternalTool(ExternalTool):
+        options_scope = "my-external-tool"
+        default_version = "1.2.3"
+        default_known_versions = [
+          "1.2.3|darwin|deadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeef|222222",
+          "1.2.3|linux |1234abcd1234abcd1234abcd1234abcd1234abcd1234abcd1234abcd1234abcd|333333",
+        ]
+
+        @classmethod
+        def generate_url(cls, plat: Platform, version: str) -> str:
+            ...
+
+        @classmethod
+        def generate_exe(cls, plat: Platform, version: str) -> str:
+            ...
+
+    @rule
+    def my_rule(my_external_tool: MyExternalTool):
+      downloaded_tool = await Get[DownloadedExternalTool](
+        ExternalToolRequest, my_external_tool.get_request(Platform.current)
+      )
+
+    ...
+    def rules():
+      return [my_rule, subsystem_rule(MyExternalTool)]
+
+
+    A lightweight replacement for the code in binary_tool.py and binary_util.py,
+    which can be deprecated in favor of this.
```

Comment:
I would like to add this as a `@deprecated_module` to those files, then -- we could potentially use `deprecation_start_version` to delay it if we expect we'll need more than one release cycle to completely remove things.

I know `binary_util.py` was also used to fetch some tools for rust bootstrapping. We'll probably want to add that before we can remove it entirely.

Reply:
Yeah, I would like to deprecate, but I'm going to transition all current uses first, to see what yelps.

Reply:
Yes, sounds great!